### PR TITLE
Upgrade to composer v2 for PHP

### PR DESCRIPTION
**Pull request by WyriHaximus**
Resolves: #2668

### CODE REVIEW

**Comment by WyriHaximus** [comment]
@jurre can you help out with the tests? Especially this one:
Screenshot from 2020-10-26 19-29-50

And alternatively, is there a way to regenerate all the lock files for the fixtures, or is that all hand work?

**Comment by WyriHaximus** [comment]
Opened a support issue for the cs fixer issue: https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/5205

**Comment by jurre** [comment]
Thanks so much for this work @WyriHaximus, really appreciated! I don't think there's currently a good way to automatically update those lockfiles, but for other package-managers we have started creating [project/repo based fixtures](), and I might have a look at migrating to that, in which case we could easily loop over the folders and run composer on all of them.

I'll try to find some time to look at the tests this week

**Comment by WyriHaximus** [comment]
No problem @jurre assuming you mean PR's like #2577 ? That would help a lot, will check if I can fix the error I mentioned here in the mean time: https://github.com/dependabot/dependabot-core/pull/2682#issuecomment-716743445

File: composer/spec/fixtures/composer_files/pear [hunk 1]
```diff
@@ -6,6 +6,6 @@
     }
   ],
   "require": {
-    "pear-pear.horde.org/Horde_Date": "^2.4.0@stable"
+    "pear-pear.horde.org/horde_date": "^2.4.0@stable"
   }
 }
```
Comment (lstrojny) [inline composer/spec/fixtures/composer_files/pear:11]: PEAR support is gone in v2

**Comment by WyriHaximus** [comment]
> Opened a support issue for the cs fixer issue: [FriendsOfPHP/PHP-CS-Fixer#5205](https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/5205)

They released a new version and updated this PR to use that release

File: composer/helpers/src/UpdateChecker.php [hunk 2]
```diff
@@ -46,7 +49,10 @@ public static function getLatestResolvableVersion(array $args): ?string
             $io->loadConfiguration($config);
         }
 
-        $installationManager = new DependabotInstallationManager();
+        $httpDownloader = Factory::createHttpDownloader($io, $config);
+        $process = new ProcessExecutor($io);
+        $loop = new Loop($httpDownloader, $process);
```
[the remark this answers is not in the document]
Comment (WyriHaximus) [reply composer/helpers/src/UpdateChecker.php:54]: Hah thanks :+1:

File: composer/spec/fixtures/composer_files/pear (hunk 1 above)
Comment (WyriHaximus) [reply composer/spec/fixtures/composer_files/pear:11]: Didn't even know that, but haven't used pear in decades. Removed the fixtures for it

**Comment by jurre** [comment]
> > Opened a support issue for the cs fixer issue: [FriendsOfPHP/PHP-CS-Fixer#5205](https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/5205)
> 
> They released a new version and updated this PR to use that release

Sweet, I've merged that fix: https://github.com/dependabot/dependabot-core/pull/2690

**Comment by WyriHaximus** [comment]
> Thanks so much for this work @WyriHaximus, really appreciated! I don't think there's currently a good way to automatically update those lockfiles, but for other package-managers we have started creating [project/repo based fixtures](https://github.com/dependabot/dependabot-core/tree/2fa251953ed47a0f2befb1edfdddb51db24f52a6/go_modules/spec/fixtures/projects), and I might have a look at migrating to that, in which case we could easily loop over the folders and run composer on all of them.
> 
> I'll try to find some time to look at the tests this week

@jurre any news on this? Had another go at this tonight and it looks like those lock files are also tightly couple to a whole bunch of tests

**Comment by jurre** [comment]
Apologies for the delay on this @WyriHaximus, it might be a few weeks before one of us can make time to dive into this as it'll be a substantial amount of work to convert all these tests to such a project structure (or alternatively manually updating all the fixtures).

**Comment by WyriHaximus** [comment]
So @jurre, the thing I noticed looking at those tests is that most of them use the same package and versions combinations to test against. It might be worth a shot mass replacing that after finding a suitable replacement set

**Comment by localheinz** [comment]
@jurre 

> Thanks so much for this work @WyriHaximus, really appreciated! I don't think there's currently a good way to automatically update those lockfiles, but for other package-managers we have started creating project/repo based fixtures, and I might have a look at migrating to that, in which case we could easily loop over the folders and run composer on all of them.
>
>I'll try to find some time to look at the tests this week

If I knew what had to be done here, I would do it.

Can you give me a hint?

**Comment by localheinz** [comment]
@jurre 

Do I understand correctly that fixtures should be organized in such a way that we have a directory for each scenario, containing all the files for a specific scenario?

**Comment by WyriHaximus** [comment]
@localheinz In a nutshell:
https://github.com/dependabot/dependabot-core/tree/main/composer/spec/fixtures/composer_files match the files in https://github.com/dependabot/dependabot-core/tree/main/composer/spec/fixtures/lockfiles

The tricky part is that those fixtures are also used in tests, so changing one test and related fixture will also affect other tests.

**Comment by localheinz** [comment]
@WyriHaximus 

Ha, right - that's why reusable fixtures really are a myth. 

🤓

**Comment by jurre** [comment]
> @jurre
> 
> Do I understand correctly that fixtures should be organized in such a way that we have a directory for each scenario, containing all the files for a specific scenario?

Yep, that's exactly the plan, I see you already made a great start in #2741, many thanks! 💛

**Comment by WyriHaximus** [comment]
@jurre Any update on improving the tests from your side? Gave it another shot today and got stuck again

**Comment by jurre** [comment]
> @jurre Any update on improving the tests from your side? Gave it another shot today and got stuck again

No, unfortunately we haven't had time to dig into this yet. Apologies about how long this is taking, your work is definitely very much appreciated.

We're planning to scope out how much work the upgrade to composer v2 (and some other ecosystem upgrades) will require this week, and then hopefully I can tell you more.

**Comment by jurre** [comment]
FYI we're building on top of this over [here](https://github.com/dependabot/dependabot-core/pull/2838) if you'd like to follow along

**Comment by WyriHaximus** [comment]
> FYI we're building on top of this over [here](https://github.com/dependabot/dependabot-core/pull/2838) if you'd like to follow along

AWESOME!!! Closing this PR in favour of #2838 :+1:

**Pull request closed without merging**